utilities/backupable/backupable_db.cc

```diff
@@ -1494,15 +1562,47 @@ Status BackupEngineImpl::AddBackupFileWorkItem(
   uint32_t checksum_value = 0;
   std::string db_id;
   std::string db_session_id;
+  // whether the checksum for a table file has been computed
+  bool has_checksum = false;
 
-  // Step 1: Prepare the relative path to destination
-  if (shared && shared_checksum) {
-    // Prepare checksum to add to file name
+  // Whenever a default checksum function name is passed in, we will verify it
+  // before copying. Note that only table files may have a known checksum name
+  // passed in.
+  //
+  // If no default checksum function name is passed in, we will calculate the
+  // checksum *before* copying in two cases (we always calcuate checksums when
+  // copying or creating for any type files):
+  // a) share_files_with_checksum is true and file type is table;
+  // b) share_table_files is true and the file exists already.
+  if (kBackupFileChecksumFuncName == src_checksum_func_name) {
     s = CalculateChecksum(src_dir + fname, db_env_, src_env_options, size_limit,
```

Comment:
Oh interesting. I just realized that we are already reading file to compute checksum before copying. (Sigh) And am I reading right that if we have to copy the file after computing that checksum, we ignore the first checksum and re-compute the checksum on copying? I think this is what you are referring to in the first HISTORY entry, but it's not clear to me.

Reply:
Right, I figured since we have the checksum before copying available, it may make sense to compare the checksum after copying against it without incurring additional costs.

Reply:
Should we make a task to remove the first checksum calculation some day? I was also under the impression that we only checksum during the streaming copy. Also it looks like this PR introduces another case where checksum would be computed prior to copy: when `kDbFileChecksumFuncName == src_checksum_func_name`. Is there a need for it or is it just there for consistency (and thus can be removed together with the other cases of computing checksum prior to copy)?

Reply:
I think some users (in an internal discussion) prefer validating the checksum in db manifest prior to copy. (Maybe we could make the validation optional.)

Reply:
> I think some users (in an internal discussion) prefer validating the checksum in db manifest prior to copy. (Maybe we could make the validation optional.)

I believe it's a flawed approach (1) to (a) read file to validate checksum against db manifest and then (b) re-read file to copy (and maybe then re-validate checksum). I argue this is much worse than (2) read for copy and validate checksum at the end of copy. Not only do you have extra I/O reads, you inject an opportunity for a transient bit flip (CPU/mem/IO) to cause havoc. The only extra cost is writing a useless copy in the case of corruption detected. If your probability of corruption is high enough to worry about that, you have serious problems.

Increasing the surface area for transient corruption without improving power to detect permanent corruption does not improve data integrity.